### PR TITLE
[GEOS-1906] DataAccessLimits serialization issue

### PR DESCRIPTION
**Pull request by fostash**
[GEOS-1906](https://osgeo-org.atlassian.net/browse/GEOS-1906)

Fix serialization of filters when access on Layer configuration page
  

# Checklist

- [x] I have read the [contribution guidelines](https://github.com/geoserver/geoserver/blob/main/CONTRIBUTING.md).
- [x] I have sent a [Contribution Licence Agreement](https://docs.geoserver.org/latest/en/developer/policies/committing.html) (not required for small changes, e.g., fixing typos in documentation).
- [x] First PR targets the `main` branch (backports managed later; ignore for branch specific issues).
- [x] All the build checks are green ([see automated QA checks](https://docs.geoserver.org/latest/en/developer/qa-guide/index.html)).

For core and extension modules:

- [x] New unit tests have been added covering the changes.
- [ ] [Documentation](https://github.com/geoserver/geoserver/tree/main/doc/en/user/source) has been updated (if change is visible to end users).
- [ ] The [REST API docs](https://github.com/geoserver/geoserver/tree/main/doc/en/api/1.0.0) have been updated (when changing configuration objects or the REST controllers).
- [ ] There is an issue in the [GeoServer Jira](https://osgeo-org.atlassian.net/browse/GEOS/summary) (except for changes that do not affect administrators or end users in any way).
- [ ] Commit message(s) must be in the form ``[GEOS-XYZWV] Title of the Jira ticket``.
- [ ] Bug fixes and small new features are presented as a single commit.
- [ ] Each commit has a single objective (if there are multiple commits, each has a separate JIRA ticket describing its goal).

### CODE REVIEW

File: src/main/src/main/java/org/geoserver/security/DataAccessLimits.java [hunk 1]
```diff
@@ -74,11 +69,7 @@ private void writeObject(ObjectOutputStream out) throws IOException {
      */
     protected void writeFilter(Filter filter, ObjectOutputStream out) throws IOException {
         if (filter != null) {
-            try (ByteArrayOutputStream bos = new ByteArrayOutputStream()) {
-                Encoder encoder = new Encoder(CONFIGURATION);
-                encoder.encode(filter, OGC.Filter, bos);
-                out.writeObject(bos.toByteArray());
-            }
+            out.writeObject(ECQL.toCQL(Collections.singletonList(filter)));
```
Comment (aaime) [inline src/main/src/main/java/org/geoserver/security/DataAccessLimits.java:72]: CQL is known to have round-tripping issues, you might serialize one filter and get back another, that is functionally equivalent, but not syntactically the same.

**Review by mprins — CHANGES_REQUESTED** [review]
this Jira [GEOS-1906](https://osgeo-org.atlassian.net/browse/GEOS-1906)  does not at all seem related to this commit

File: src/main/src/main/java/org/geoserver/security/DataAccessLimits.java (hunk 1 above)
Comment (fostash) [reply src/main/src/main/java/org/geoserver/security/DataAccessLimits.java]: I changed it using the other method that accept a single Filter. is that what you was pointing to? or I should not use the `toCQL` method?

Comment (aaime) [reply src/main/src/main/java/org/geoserver/security/DataAccessLimits.java]: Best not to use CQL for filter serialization at all, due to the round-tripping issues I was referring to above.

File: src/main/src/main/java/org/geoserver/security/DataAccessLimits.java [hunk 2]
```diff
@@ -64,12 +69,20 @@ private void writeObject(ObjectOutputStream out) throws IOException {
     }
 
     /**
-     * Writes the non Serializable Filter object to the ObjectOutputStream using ECQL
+     * Writes the non Serializable Filter object ot the ObjectOutputStream via a OGC Filter XML
```
Comment (bradh) [inline src/main/src/main/java/org/geoserver/security/DataAccessLimits.java:72]: typo introduced.

**Comment by fostash** [comment]
you're right. I will change the commit ones I have the correct JIRA ticket

File: src/main/src/main/java/org/geoserver/security/DataAccessLimits.java [hunk 3]
```diff
@@ -69,15 +69,19 @@ private void writeObject(ObjectOutputStream out) throws IOException {
...
-                encoder.encode(filter, OGC.Filter, bos);
-                out.writeObject(bos.toByteArray());
+            if (filter != Filter.INCLUDE && filter != Filter.EXCLUDE) {
+                try (ByteArrayOutputStream bos = new ByteArrayOutputStream()) {
+                    Encoder encoder = new Encoder(CONFIGURATION);
+                    encoder.encode(filter, OGC.Filter, bos);
+                    out.writeObject(bos.toByteArray());
+                }
+            } else {
+                out.writeObject(filter);
```
Comment (aaime) [inline src/main/src/main/java/org/geoserver/security/DataAccessLimits.java:84]: Well well, I did not know INCLUDE and EXCLUDE were serializable... nice find.

**Review by aaime — APPROVED** [review]

**Comment by aaime** [comment]
I've created a Jira ticket for you, https://osgeo-org.atlassian.net/browse/GEOS-10470, and squash merged.
Backport as you see fit.

**Pull request merged**